### PR TITLE
Fix more spectral norm bugs

### PR DESCRIPTION
**Pull request by ssnl**
Problems with SN and DP after #12671 : 
1. in eval mode, `weight_orig` is not getting correct gradient #12737 .

    Fix: keep `v` vector around as a buffer and always calculate `W = W_orig / (u @ W_orig @ v)` even in eval.

2. in training mode, the `weight` buffer of the parallelized module is never updated, if someone touches `weight_orig` and/or `weight` and makes them not sharing storage. So in `eval` the weight used is wrong.

    Fix: Make `weight` not a buffer anymore and always calculate it as above.

3. #12671 changed SN to update `u` in-place to make DP work correctly, but then it breaks backward through two forwards (e.g., the common GAN loss `D(real) - D(fake)`) because the vectors needed to backprop the 1st forward is changed in the 2nd forward. 

    Fix: This PR clones `u` and `v` before using them.

To maintain BC, I added a hook interface for producing and loading state_dict. This is ugly and we should really have better interface for spectral_norm. But for the purpose to fix this issue, I make this patch. Even if we have a better interface, BC mechanism for legacy loading legacy state_dict still needs to be done.

cc @t-vi @crcrpar

### CODE REVIEW

**Comment by SsnL** [comment]
@YaoshengFu This PR will fix the spectral norm bug you see.

File: torch/nn/utils/spectral_norm.py [hunk 1]
```diff
@@ -17,90 +25,168 @@ def __init__(self, name='weight', n_power_iterations=1, dim=0, eps=1e-12):
...
+        # (the invariant at top of this class) and `u @ W @ v = sigma`.
+        # This uses pinverse in case W^T W is not invertible.
+        v = torch.chain_matmul(weight_mat.t().mm(weight_mat).pinverse(), weight_mat.t(), u.unsqueeze(1)).squeeze(1)
+        return v.mul_(target_sigma / torch.dot(u, torch.mv(weight_mat, v)))
 
     @staticmethod
     def apply(module, name, n_power_iterations, dim, eps):
+        for k, hook in module._forward_pre_hooks.items():
+            if isinstance(hook, SpectralNorm) and hook.name == name:
+                raise RuntimeError("Cannot register two spectral_norm hooks on "
```
Comment (crcrpar) [inline torch/nn/utils/spectral_norm.py]: Thanks for adding this.

**Comment by crcrpar** [comment]
I really always appreciate you guys.
 
I have one question as to my understanding.
> `u` and `v` can/should be updated in a in-palce manner while `weight_orig` not.
> This is why `weight_orig` is registered as an attribute.  

Is this correct?

**Comment by SsnL** [comment]
@crcrpar Almost! `weight_orig` is still a parameter, so updates to it should be done on the parallelized module (the one you pass over to `DataParallel`) via things like optimizers explicitly started by user. Even if it is not updated in-place, it is fine, because the next time `DataParallel` copies it to the devices, the new `weight_orig` will be get from the parallelized module, and broadcast over. 

`u` and `v`, however, are a bit different because they are updated when the module is activated, and not within users' control. So it is our job to automatically update it. And to ensure that such update works with `DataParallel`, it needs to be done in-place.

File: torch/nn/modules/module.py [hunk 2]
```diff
@@ -66,6 +66,8 @@ def __init__(self):
         self._backward_hooks = OrderedDict()
         self._forward_hooks = OrderedDict()
         self._forward_pre_hooks = OrderedDict()
+        self._state_dict_hooks = OrderedDict()
```
Comment (colesbury) [inline torch/nn/modules/module.py]: I think you may need to conditionally create these in `__setstate__`

**Review by colesbury — APPROVED** [review]
lgtm, but I'm not very familiar with the spectral norm code feel a bit lost.

If @t-vi has time and is familiar with this, his review may be helpful.

File: torch/nn/modules/module.py (hunk 2 above)
Comment (SsnL) [reply torch/nn/modules/module.py]: Thanks! Fixed.

**Comment by SsnL** [comment]
@t-vi and @crcrpar It would be great if one or both of you could take a look at the SN changes :)

File: test/test_nn.py [hunk 3]
```diff
@@ -1749,58 +1754,173 @@ def test_spectral_norm(self):
...
+                pre_remove_out = wrapped_m(input)
+                m = torch.nn.utils.remove_spectral_norm(m)
+                self.assertEqual(wrapped_m(input), pre_remove_out)
+
+                m = torch.nn.utils.spectral_norm(m)
+                m = torch.nn.utils.remove_spectral_norm(m)
+                self.assertEqual(wrapped_m(input), pre_remove_out)
+
+                m = torch.nn.utils.spectral_norm(m)
+                for i in range(3):
```
Comment (crcrpar) [inline test/test_nn.py]: nit: Is 3 selected randomly?

File: test/test_nn.py [hunk 4]
```diff
@@ -1749,58 +1754,173 @@ def test_spectral_norm(self):
...
+                    pre_remove_out = wrapped_m(input)
+                m = torch.nn.utils.remove_spectral_norm(m)
+                self.assertEqual(wrapped_m(input), pre_remove_out)
+
+                # TEST EVAL BEHAVIOR
+
+                m = torch.nn.utils.spectral_norm(m)
+                wrapped_m(input)
```
Comment (crcrpar) [inline test/test_nn.py]: nit: Is this forward necessary?

**Comment by crcrpar** [comment]
Sorry for my late response.

Changes are good but I have some questions about test codes.
So could you tell me if you have the time?

File: test/test_nn.py (hunk 3 above)
Comment (SsnL) [reply test/test_nn.py]: yes, randomly :)

File: test/test_nn.py (hunk 4 above)
Comment (SsnL) [reply test/test_nn.py]: not really, but I thought testing eval after training more than 1 time would be more comprehensive.

**Comment by SsnL** [comment]
Thank you for your comments @crcrpar :)

File: test/test_nn.py (hunk 4 above)
Comment (crcrpar) [reply test/test_nn.py]: year, that makes sense to me.

File: test/test_nn.py (hunk 3 above)
Comment (crcrpar) [reply test/test_nn.py]: thanks

**Comment by zmurez** [comment]
I am still having a problem with spectral norm complaining about in place changes when followed by batch normalization and distributed across 2 or more GPUs using data parallel. This architecture arises in Big GAN (https://arxiv.org/abs/1809.11096). Interestingly this is not an issue with group normalization, instance normalization, or no normalization. It is also not an issue on a single GPU.

**Comment by t-vi** [comment]
So to get this a bit sorted, how many cases do we have? with/without DP * eval/training * weight.requires_grad=True/False * ???.
I'd feel much more comfortable if I had a list of everything that is expected to work.

**Comment by SsnL** [comment]
@zmurez even with this patch?

**Comment by SsnL** [comment]
@t-vi Yeah, I think those 8 cases are all we need to consider.

**Comment by zmurez** [comment]
> @zmurez even with this patch?

I think so. It complained regardless of normalization and number of GPU's prior to adding this patch. However, I just copied the relevant lines into my own spectral_norm.py file instead of pulling the entire branch... So it is possible I missed something... but it seems unlikely since all the other cases are fixed.

**Comment by SsnL** [comment]
@zmurez It is possible that other changes are needed. Do you have a small repro script I can try on my build?

**Comment by zmurez** [comment]
> @zmurez It is possible that other changes are needed. Do you have a small repro script I can try on my build?

```
import torch
import torch.nn as nn

#from spectral_norm import spectral_norm       # my local spectral_norm patch
from torch.nn.utils import spectral_norm       # torch implementation

dim=5
batchsize=10
net = nn.DataParallel(nn.Sequential(
            #nn.Conv2d(dim, dim,1),
            nn.BatchNorm2d(dim),
            #nn.LeakyReLU(0.2, inplace=True),
            spectral_norm(nn.Conv2d(dim, dim,1)),
        )).cuda()
noise = torch.randn(batchsize, dim, 1, 1).cuda()
net.zero_grad()
out = net(noise).sum()
out.backward()
print('no bug')

**Comment by SsnL** [comment]
@zmurez Thanks! This is actually a bug elsewhere. I have submitted a fix at https://github.com/pytorch/pytorch/pull/13594. With these two patches together, I can run your script successfully.

File: torch/nn/utils/spectral_norm.py [hunk 5]
```diff
@@ -17,90 +25,168 @@ def __init__(self, name='weight', n_power_iterations=1, dim=0, eps=1e-12):
...
+        fn = self.fn
+        version = local_metadata.get('spectral_norm', {}).get(fn.name + '.version', None)
+        if version is None or version < 1:
+            with torch.no_grad():
+                weight_orig = state_dict[prefix + fn.name + '_orig']
+                weight = state_dict.pop(prefix + fn.name)
+                sigma = (weight_orig / weight).mean()
+                weight_mat = fn.reshape_weight_to_matrix(weight_orig)
+                u = state_dict[prefix + fn.name + '_u']
+                v = fn._solve_v_and_rescale(weight_mat, u, sigma)
```
Comment (t-vi) [inline torch/nn/utils/spectral_norm.py]: Is it OK to have computation during loading the state dict? I think it probably is, but it should be a conscious decision.

**Review by t-vi — COMMENTED** [review]
Thanks for adding all the new tests! They seem to be comprehensive and I think the patch is good with them.

File: torch/nn/utils/spectral_norm.py (hunk 5 above)
Comment (SsnL) [reply torch/nn/utils/spectral_norm.py]: Yes it is intended. Old state dict don't have `v` so we have to "solve" for it somehow. It can't be after loading the state dict because we lose info about the version in the state dict by then.

**Comment by SsnL** [comment]
@t-vi @crcrpar Thank you for your reviews! I appreciate them. :)

**Comment by zmurez** [comment]
It seems that when spectral norm is applied to a conv module, u is randomly initialized. Then v is computed such that the invariant holds.

1) This initial solve is very very slow (over tens of minutes) for large networks.
a) Is this really necessary or is a random initialization ok?
b) Can this solve be replaced by a sequence of power iterations?

2) If the weights of the conv module are later initialized using nn.init, I think this invariant will be broken?

**Comment by SsnL** [comment]
@zmurez Good point. I'll think about ways to fix it.

**Comment by t-vi** [comment]
For the initialization I think it should be not terribly important whether to init u or v randomly, so that could likely be changed. For loading the state dict: this only happens when there you load an "old version" state dict. How about warning about the performance and (potentially) offering a switch to forward re-compute u instead of solving for v. (I think it might be save to "do the right, slow thing" by default rather than the other way round.)

**Comment by SsnL** [comment]
@t-vi @crcrpar @zmurez Thank you. I have removed the solving part in the new commit :)
